sweagent/agent/models.py

```diff
@@ -565,15 +566,17 @@ def query(self, history: list[dict[str, str]]) -> str:
         completion = together.Complete.create(
             model=self.api_model,
             prompt=prompt,
-            max_tokens=self.model_metadata["max_context"],
-            stop="<human>",
+            # Input token count + max_tokens needs to be less than the context length of the model.
+            # But it is difficult to calculate input token before request.
+            max_tokens=None,
+            stop=["<human>"],
```

Comment:
here is new request definitions.
https://docs.together.ai/reference/complete

Reply:
Thanks for your great review! I misunderstood a bit.

> It seems to me that max_tokens sets the maximum number of output tokens, though, right?

Yes. But it seems that `max_token` is different from `context_length`. We need to specify `max_tokens` as the value obtained by subtracting the input tokens from `context_length`.

[This link](https://github.com/princeton-nlp/SWE-agent/issues/135#issuecomment-2054060266) is the error message when we set `max_tokens` as `max_context` in MODELS constants.

So, it seems better to calculate `input_token` in advance like [Anthropics' Claude-2 models](https://github.com/mikanfactory/SWE-agent/blob/main/sweagent/agent/models.py#L389) even though it's may not be accurate.

I have a question about the implementation. Which one of the following, 1 or 2, would be better?

1. Like Anthropic's Claude-2 models, use the `count_token` method from the Anthropic package.
    - While this implementation will be quick, it feels unnatural to rely on the Anthropic package when using the Together model.
2. Imitate the `count_token` method from the Anthropic package and implement it within the SWE-agent.
    - Looking back from [here](https://github.com/anthropics/anthropic-sdk-python/blob/8e3d8a68d309424238ae54e03ee962f7147cfc60/src/anthropic/_client.py#L270), it seems that we are only using the huggingface/tokenizer (and it as a singleton).

I think 2 is better because 1 feels unnatural, but I would appreciate advice on other approaches as well!

Reply:
It feels a bit weird, but I implemented 1.